pinot-spi/src/main/java/org/apache/pinot/spi/environmentprovider/PinotEnvironmentProviderFactory.java

```diff
@@ -0,0 +1,82 @@
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pinot.spi.environmentprovider;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.Iterables;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.apache.pinot.spi.plugin.PluginManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This factory class initializes the PinotEnvironmentProvider class.
+ * It creates a PinotEnvironment object based on the URI found.
+ */
+public class PinotEnvironmentProviderFactory {
+
+  private PinotEnvironmentProviderFactory() {
+  }
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(PinotEnvironmentProviderFactory.class);
+  private static final String CLASS = "class";
+  private static final Map<String, PinotEnvironmentProvider> PINOT_ENVIRONMENT_PROVIDER_MAP = new HashMap<>();
+
+  public static void init(PinotConfiguration environmentProviderFactoryConfig) {
+    // Get environment and it's respective class
+    PinotConfiguration environmentConfiguration = environmentProviderFactoryConfig.subset(CLASS);
+    List<String> environments = environmentConfiguration.getKeys();
+
+    if (environments.isEmpty()) {
+      LOGGER.info("Did not find any environment provider classes in the configuration");
+      return;
+    }
+
+    String environment = Iterables.getOnlyElement(environments);
+    String environmentProviderClassName = environmentConfiguration.getProperty(environment);
+    PinotConfiguration environmentProviderConfiguration = environmentProviderFactoryConfig.subset(environment);
+    LOGGER.info("Got environment {}, initializing class {}", environment, environmentProviderClassName);
+    register(environment, environmentProviderClassName, environmentProviderConfiguration);
+  }
+
+  // Utility to invoke the cloud specific environment provider.
+  public static PinotEnvironmentProvider getEnvironmentProvider(String environment) {
+    PinotEnvironmentProvider pinotEnvironmentProvider = PINOT_ENVIRONMENT_PROVIDER_MAP.get(environment);
+    Preconditions.checkState(pinotEnvironmentProvider != null,
+        "PinotEnvironmentProvider for environment: %s has not been initialized", environment);
+    return pinotEnvironmentProvider;
+  }
+
+  private static void register(
+      String environment, String environmentProviderFileName, PinotConfiguration environmentProviderConfiguration) {
+    try {
+      LOGGER.info("Initializing PinotEnvironmentProvider for environment {}, classname {}", environment, environmentProviderFileName);
+      PinotEnvironmentProvider pinotEnvironmentProvider = PluginManager.get().createInstance(environmentProviderFileName);
+      pinotEnvironmentProvider.init(environmentProviderConfiguration);
+      PINOT_ENVIRONMENT_PROVIDER_MAP.put(environment, pinotEnvironmentProvider);
```

Comment:
It may be a good idea to at least detect duplicate registrations and register either the first or last one (could be useful in an Integration test env where we may start more than one server, for example)